Make `has_unitary` and `unitary` fallback to `decompose`

andbe91 opened the pull request:
Intended to fix #1163 
So far, I have made `has_unitary` and `unitary` fallback to  `decompose`, together with some tests of the new functionality. Note that only objects that are instances of `cirq.Operation` are decomposed at the moment.
There is currently one test failing, which I have yet to solve (I did have 704 failed tests at some point, so progress!). 

@Strilanc If I could get some comment on what I have done so far, what more tests I should implement, and "As part of this issue, a lot of special case "decompose to get unitary" code strewn around the code base (e.g. in simulators) should be rewritten.", that would be very much appreciated.

Strilanc commented on cirq/protocols/unitary_test.py:
```diff
@@ -74,3 +101,33 @@ def _unitary_(self) -> np.ndarray:
+    np.testing.assert_allclose(cirq.unitary(cirq.GateOperation(
+        DecomposableRecursion(True), (a,))), m)
+
+    np.testing.assert_allclose(cirq.unitary(cirq.GateOperation(
```
Make sure to exercise the case where sub-operations apply to a subset of qubits, or to qubits in a different order.

Strilanc commented on cirq/protocols/unitary_test.py:
```diff
@@ -74,3 +101,33 @@ def _unitary_(self) -> np.ndarray:
+    np.testing.assert_allclose(cirq.unitary(cirq.GateOperation(
+        DecomposableRecursion(True), (a, b))), m2)
+
+    # assert cirq.unitary(FullyImplemented(False), None)
```
commented line

andbe91 commented on cirq/protocols/unitary.py:
```diff
@@ -109,13 +139,22 @@ def has_unitary(val: Any) -> bool:
     result = NotImplemented if getter is None else getter()
 
+    # Fallback to decomposition for operations
+    if (result is NotImplemented or not result) and isinstance(val, Operation):
```
Just to make sure, you don't want the `if result is False`?
This causes an issue in the following way if `val` is a `GateOperation`:

1. We first check if `val` has the `_has_unitary_` method, which all instances of `GateOperation` do.
2. `GateOperation._has_unitary_` returns `has_unitary(self._gate)`.
3. If `_gate._has_unitary_` is not implemented, it will not try to decompose it since it is not a `GateOperation` anymore, and will therefore return `False`.
4. Therefore, we will not enter the decompose branch if we don't include  `if result is False`.

It might come down to this, does `has_unitary(val) == False` mean that a unitary doesn't exist, or that we just don't know what it is?

Strilanc replied:
Hmmm, I see. The fact that we're limiting ourselves to just operation for now is causing gate operation to not work, because it's implemented in terms of the gate.

I think the correct move here is to trust the False instead of trying to override it. It will start to work once we add support for gates. Maybe that indicates we should do it as part of this PR.

andbe91 commented on cirq/protocols/unitary.py:
```diff
@@ -109,13 +139,22 @@ def has_unitary(val: Any) -> bool:
+    # Fallback to decomposition for operations
+    if (result is NotImplemented or not result) and isinstance(val, Operation):
+        decomposed_val = decompose_once(val, [])
+        if (len(decomposed_val) > 0 and
```
Let's say that an object has neither `_has_unitary_` or `_unitary_` defined, but `_decompose_` returns an empty list. Then according to your comment, `unitary` will return True.
If I then call `unitary` on the same object, what should it return?

An example of this case is the `DummyOperation` in qasm_output_test.py.

andbe91 replied:
I realized that this of course should just be the identity matrix. That solved some of my issues.

andbe91 commented:
I hope that I addressed all your comments on my last commit.

About the failing test in convert_to_cz_and_single_gates_test.py, it tries to optimize a two qubit circuit consisting of composite gates. Before these composite gates did not have unitaries (the name of the test is also test_composite_gates_without_matrix), but with the new implementation of unitary and has_unitary they do now. This seems to lead to a different global phase for the optimized circuit since
 `cirq.testing.assert_allclose_up_to_global_phase(
        circuit.to_unitary_matrix(),
        c_orig.to_unitary_matrix(),
        atol=1e-7)`
passes fine, but `assert circuit == expected` does not.

If one indeed prints the two circuits, 
circuit:
0: ───X───Y^0.5───@───────────────
                  │
1: ───────────────@───Y^0.5───Z───

expected:
0: ───X───Y^0.5───@───────────────
                  │
1: ───────────────@───X───Y^0.5───

Is this test no longer meaningful, is there something wrong, or how should I handle this?

andbe91 commented:
Hi @Strilanc, could you please review this again? It should now work correctly for operations, but not yet for gates. Should I try to implement this for gates as part of this PR? If so, what would be the best way to pass the qubit order to `unitary`?

andbe91 commented:
I think that the `has_unitary` part should be done (it checks if all decomposed objects has unitaries), am I missing some functionality here?

I had some new tests from controlled operation fail, I tracked it down to that it was trying to get the unitary of a parameterized CCZ gate. What happens is that `CCZPowGate._unitary_()` returns `NotImplemented` if `is_parameterized` returns `True`, this leads to that `cirq.unitary` tries to decompose the gate, however `CCZPowGate._decompose_` raises an error if `CCZPowGate._exponent` is a symbol. To solve this I found three different solutions:
1. Make `CCZPowGate._decompose_` return NotImplemented if it is parameterized.
2. Make `cirq.unitary`  (`cirq.has_unitary`) check if `val` is parameterized and return default (False) if it is.
3. Only decompose in `cirq.unitary` ( `cirq.has_unitary`) if `val._unitary` (`val._has_unitary`) doesn't exist, i.e. if they do exist and returns NotImplemented, we _don't_ decompose.

Which of these alternatives would you prefer @Strilanc?

Strilanc commented:
Make `CCZPowGate._decompose_` return `NotImplemented` if it is parameterized.

Strilanc commented:
To be specific:

```python
    def _decompose_(self, qubits):
        """An adjacency-respecting decomposition.

        0: ───p───@──────────────@───────@──────────@──────────
                  │              │       │          │
        1: ───p───X───@───p^-1───X───@───X──────@───X──────@───
                      │              │          │          │
        2: ───p───────X───p──────────X───p^-1───X───p^-1───X───

        where p = T**self._exponent
        """
        if protocols.is_parameterized(self):
            return NotImplemented

        a, b, c = qubits

        # Hacky magic: avoid the non-adjacent edge.
        if hasattr(b, 'is_adjacent'):
            if not b.is_adjacent(a):
                b, c = c, b
            elif not b.is_adjacent(c):
                a, b = b, a

        p = common_gates.T**self._exponent
        sweep_abc = [common_gates.CNOT(a, b),
                     common_gates.CNOT(b, c)]

        return [
            p(a), p(b), p(c),
            sweep_abc,
            p(b)**-1, p(c),
            sweep_abc,
            p(c)**-1,
            sweep_abc,
            p(c)**-1,
            sweep_abc,
        ]
```

Strilanc commented:
You can ignore the `num_qubits` method coverages via `# coverage: ignore`

andbe91 commented:
Thanks for the help! This should finally be ready to merge.
